plonk/src/circuit/basic.rs

```diff
@@ -1416,6 +1488,105 @@ impl<F: PrimeField> PlonkCircuit<F> {
     }
 }
 
+/// Private helper functions for comparison gate
+impl<F: PrimeField> PlonkCircuit<F> {
+    /// Return a variable indicating whether `a` <= `b`.
+    fn is_leq_internal(&mut self, a: Variable, b: Variable) -> Result<BoolVar, PlonkError> {
+        let a_le_const =
+            self.is_leq_constant_internal(a, &F::from(F::modulus_minus_one_div_two()))?;
+        let b_le_const =
+            self.is_leq_constant_internal(b, &F::from(F::modulus_minus_one_div_two()))?;
+        let b_greater_const = self.logic_neg(b_le_const)?;
+        // Check whether `a` <= (q-1)/2 and `b` > (q-1)/2
+        let msb_check = self.logic_and(a_le_const, b_greater_const)?;
+
+        // Check whether `a` and `b` are both <= (q-1)/2 or
+        // are both > (q-1)/2
+        let msb_eq = self.check_equal(a_le_const.0, b_le_const.0)?;
+        // If so, check whether (b-a) <= (q-1)/2
+        let c = self.sub(b, a)?;
+        let cmp_result =
+            self.is_leq_constant_internal(c, &F::from(F::modulus_minus_one_div_two()))?;
+        let cmp_result = self.logic_and(msb_eq, cmp_result)?;
+
+        self.logic_or(msb_check, cmp_result)
+    }
+
+    /// Constrain that `a` <= `b`
+    fn enforce_leq_internal(&mut self, a: Variable, b: Variable) -> Result<(), PlonkError> {
+        let a_le_const =
+            self.is_leq_constant_internal(a, &F::from(F::modulus_minus_one_div_two()))?;
+        let b_le_const =
+            self.is_leq_constant_internal(b, &F::from(F::modulus_minus_one_div_two()))?;
+        let b_ge_const = self.logic_neg(b_le_const)?;
+        // Check whether `a` <= (q-1)/2 and `b` > (q-1)/2
+        let msb_check = self.logic_and(a_le_const, b_ge_const)?;
+
+        // Check whether `a` and `b` are both <= (q-1)/2 or
+        // are both > (q-1)/2
+        let msb_eq = self.check_equal(a_le_const.0, b_le_const.0)?;
+        // If so, check whether (b-a) <= (q-1)/2
+        let c = self.sub(b, a)?;
+        let cmp_result =
+            self.is_leq_constant_internal(c, &F::from(F::modulus_minus_one_div_two()))?;
+        let cmp_result = self.logic_and(msb_eq, cmp_result)?;
+
+        self.logic_or_gate(msb_check, cmp_result)
+    }
+
+    /// Helper function to check whether `a` is no greater
+    /// than a given constant. Let N = F::size_in_bits()
+    /// It uses N negation gates and N AND/OR gates.
+    fn is_leq_constant_internal(
+        &mut self,
+        a: Variable,
+        constant: &F,
+    ) -> Result<BoolVar, PlonkError> {
+        let a_bits_le = self.unpack(a, F::size_in_bits())?;
```

Comment:
If we already know that all bit variables are bool, can we use a1 + ... + an = 0 to constrain that all {ai} are zeroes? This will save many gates below.

Reply:
Was thinking about it but not sure.